services/workflows-service/src/alert/alert.service.ts

```diff
@@ -103,23 +113,132 @@ export class AlertService {
+      return false;
+    }
+
+    // Find the missing keys
+    results.forEach((aggreatedRow: any) => {
+      const missingKeys = _.difference(inlineRule.groupedBy, Object.keys(aggreatedRow));
+      // If there are missing keys, log an error message
+      if (missingKeys.length > 0) {
+        console.error(
+          `Alert aggregated row is missing properties for groupBy: ${missingKeys.join(', ')}`,
+          {
+            inlineRule,
+            aggreatedRow,
+          },
+        );
+      }
+    });
+
+    // const releavntAlerts = await this.getDeduplicatedAlerts(
+    //   alertDefinition.projectId,
+    //   alertDefinition,
+    //   results,
+    // );
+
+    const alertsSetteledResult = await Promise.allSettled<
+      { alert: Alert; alertDefinition: AlertDefinition }[]
+    >(
+      results.map(async (aggreatedRow: any) => {
+        const data = _.pick(aggreatedRow, inlineRule.groupedBy);
+
+        return {
+          alertDefinition,
+          alert: await this.createAlert(alertDefinition, data),
+        };
+      }),
+    );
+
+    const result = {
+      fulfilled: [],
+      rejected: [],
+    } as any;
+
+    alertsSetteledResult.forEach(resultItem => {
+      if (resultItem.status === 'fulfilled') {
+        result.fulfilled.push(resultItem.value);
+      } else if (resultItem.status === 'rejected') {
+        result.rejected.push(resultItem.reason);
+      }
+    });
+
+    return result;
+  }
+
+  private createAlert(
+    alertDef: AlertDefinition,
+    data: Pick<Prisma.AlertUncheckedCreateInput, 'businessId' | 'endUserId' | 'counterpartyId'>,
+  ): Promise<Alert> {
+    return this.alertRepository.create({
+      data: {
+        alertDefinitionId: alertDef.id,
+        projectId: alertDef.projectId,
+        severity: alertDef.defaultSeverity,
+        dataTimestamp: new Date(),
+        state: AlertState.Triggered,
+        status: AlertStatus.New,
+        executionDetails: {},
+        ...data,
+      },
+    });
+  }
+
+  private async getDeduplicatedAlerts(
+    projectId: string,
+    alertDef: AlertDefinition,
+    data: any,
+  ): Promise<any> {
+    const cooldownTimeframe = (alertDef.dedupeStrategies as any).cooldownTimeframe;
+
+    const potentialDuplicates = await this.alertRepository.findMany(
+      {
+        where: {
+          AND: [
+            { projectId: projectId },
+            { alertDefinitionId: alertDef.id },
+            { dataTimestamp: { gte: new Date(data.dataTimestamp - cooldownTimeframe) } },
+            // Adjust the cooldown timeframe as needed
+          ],
+        },
+      },
+      [projectId],
+    );
+
+    const groupedDuplicates = potentialDuplicates.reduce((acc: any, alert) => {
+      const key = `${alert.projectId}_${alert.alertDefinitionId}_${Math.floor(
+        alert.dataTimestamp.getTime() / cooldownTimeframe,
+      )}`;
+      acc[key] = acc[key] || [];
+      acc[key].push(alert);
+      return acc;
+    }, {});
```

Comment:
`Array.prototype.reduce` expects type to be passed as a type generic `reduce<...>` or by using `as` on the second argument i.e. `{} as ...`. The `as` way tends to be the way to go.